src/client.rs

```diff
@@ -21,56 +26,124 @@ pub enum Client {
     Socks5(RawClient<ElectrumProxyStream>),
 }
 
+/// Generalized Electrum client that supports multiple backends. Can re-instantiate client_type if connections
+/// drops
+pub struct Client {
+    client_type: RwLock<ClientType>,
+    config: Config,
+    url: String,
+}
+
 macro_rules! impl_inner_call {
     ( $self:expr, $name:ident $(, $args:expr)* ) => {
-        match $self {
-            Client::TCP(inner) => inner.$name( $($args, )* ),
-            Client::SSL(inner) => inner.$name( $($args, )* ),
-            Client::Socks5(inner) => inner.$name( $($args, )* ),
+    {
+        let mut errors = vec![];
+        loop {
+            let read_client = $self.client_type.read().unwrap();
+            let res = match &*read_client {
+                ClientType::TCP(inner) => inner.$name( $($args, )* ),
+                ClientType::SSL(inner) => inner.$name( $($args, )* ),
+                ClientType::Socks5(inner) => inner.$name( $($args, )* ),
+            };
+            drop(read_client);
+            match res {
+                Ok(val) => return Ok(val),
+                Err(Error::Protocol(_)) => {
+                    return res;
+                },
+                Err(e) => {
+                    warn!("call retry:{}/{} {:?}", errors.len() + 1 , $self.config.retry(), e);
+                    errors.push(e);
+                    if errors.len() as u8 == $self.config.retry() {
+                        return Err(Error::AllAttemptsErrored(errors));
+                    }
+
+                    // Only one thread will try to recreate the client getting the write lock,
+                    // other eventual threads will get Err and will block at the beginning of
+                    // previous loop when trying to read()
+                    if let Ok(mut write_client) = $self.client_type.try_write() {
+                        loop {
+                            std::thread::sleep(std::time::Duration::from_secs(errors.len() as u64));
```

Comment:
Do you think it's better to keep this linear or switch to an exponential backoff, potentially with a cap to a reasonable value like 16 seconds or something like that?

Reply:
what do you think of something like `(errors.len() * errors.len()).min(60)`

Reply:
I was thinking about `(1 << errors.len()).min(something)` but that should also work. Not sure what's the best really ahaha

Also not sure what's a good value for the cap: 60 seconds between two tries seems a bit high to me, maybe around 30 is better?

Reply:
Ok, now `(1 << errors.len()).min(30)`